File: Graphics/imcmp.pl
#: imcmp.pl
#: Compare two (very simple) images
#: Copyright (c) 2006 Agent Zhang
#: 2006-03-17 2006-04-17

use strict;
use warnings;
use List::Util qw(min);
use GD;
use Text::Table;

my ($file1, $file2) = @ARGV;
if (!$file1 || !$file2) {
    die "Usage: imcmp <image1> <image2>\n";
}

my $tb = Text::Table->new(
    "Coordinates", $file1, $file2,
);

my $im1 = GD::Image->new($file1) or die "$!";
my $im2 = GD::Image->new($file2) or die "$!";

my ($w1, $h1) = $im1->getBounds();
my ($w2, $h2) = $im2->getBounds();

my $w = min($w1, $w2);
my $h = min($h1, $h2);

my $diff = 0;
for my $x (0..$w-1) {
    for my $y (0..$h-1) {
        my $c1 = get_color($im1, $x, $y);
        my $c2 = get_color($im2, $x, $y);
        if ($c1 ne $c2) {
            $tb->add("($x, $y)", $c1, $c2);
            $diff = 1;
        }
    }
}

if ($diff) {
    print $tb->rule( '-', '+' );
    print $tb->title;
    print $tb->rule( '-', '+' );
    print $tb->body;
    print $tb->rule( '-', '+' );
}

sub get_color {
    my ($im, $x, $y) = @_;
    my $index = $im->getPixel($x, $y);
    my ($r, $g, $b) = $im->rgb($index);
    my $c = rgb2name($r, $g, $b);
    return $c;
}

sub rgb2name {
    my ($r, $g, $b) = @_;
    my $c = "<$r, $g, $b>";
    if ($c eq '<255, 255, 255>') {
        return 'white';
    } elsif ($c eq '<255, 0, 0>') {
        return 'red';
    } elsif ($c eq '<0, 0, 0>') {
        return 'black';
    } elsif ($c eq '<0, 255, 0>') {
        return 'green';
    } elsif ($c eq '<0, 0, 255>') {
        return 'blue';
    } else {
        return "<$r, $g, $b>";
    }
}

File: Graphics/GD++/t/imcmp.cpp
//: imcmp.cpp
//: Copyright (c) 2006 Agent Zhang
//: 2006-04-17 2006-04-17

#include <string>
#include <stdio.h>
#include <gdpp.h>

using namespace std;

string rgb2name (int R, int G, int B) {
    if (R == 255 && G == 255 && B == 255) {
        return "white";
    }
    if (R == 255 && G == 0 && B == 0) {
        return "red";
    }
    if (R == 0 && G == 0 && B == 0) {
        return "black";
    }
    if (R == 0 && G == 255 && B == 0) {
        return "green";
    }
    if (R == 0 && G == 0 && B == 255) {
        return "blue";
    }
    return "<" + GD::Simple::itos(R) + ", " +
        GD::Simple::itos(G) + ", " +
        GD::Simple::itos(B) + ">";
}

string get_color(GD::Simple im, int x, int y) {
    GD::Color c = im.getPixel(x, y);
    return rgb2name(c.R(), c.G(), c.B());
}

int main(int argc, char* argv[]) {
    if (argc != 2+1) {
        fprintf(stderr, "Usage: imcmp <image1> <image2>\n");
        return 2;
    }

    string file1 = argv[1];
    string file2 = argv[2];

    printf("Coordinates\t\t%s\t\t%s\n", file1.c_str(), file2.c_str());

    GD::Simple im1(file1);
    GD::Simple im2(file2);

    int w1 = im1.width();
    int h1 = im1.height();
    int w2, h2;
    im2.getBounds(w2, h2);

    if (w1 != w2 || h1 != h2) {
        fprintf(
            stderr, 
            "Images %s and %s are of different sizes.\n",
            file1.c_str(), file2.c_str()
        );
        return 1;
    }

    bool diff = false;
    for (int x = 0; x < w1; x++) {
        for (int y = 0; y < h1; y++) {
            string c1 = get_color(im1, x, y);
            string c2 = get_color(im2, x, y);
            if (c1 != c2) {
                printf("(%d, %d)\t\t%s\t\t%s\n",
                    x, y, c1.c_str(), c2.c_str()
                );
                diff = true;
            }
        }
    }

    return diff ? 1 : 0;
}
